src/lib.rs

```diff
@@ -36,13 +38,10 @@ pub struct TodoOperation {
     todo: TodoData,
 }
 
-pub trait ToOperation {
-    fn to_operation_string(&self) -> String;
-}
-
-impl ToOperation for TodoOperation {
-    fn to_operation_string(&self) -> String {
-        format!(
+impl Display for TodoOperation {
```

Comment:
Implementing `Display` gives us `to_string` for free, and it also skips an allocation of an intermediate `String`, which is nice: it tells Rust *how* to format it for any given formatter, but it doesn’t require allocating until whatever point it actually needs it.

Reply:
I thought about this initially, but I thought that what I was doing was misaligned with what Display was for. The delimited format is _only_ for storage, and never for human representation. Specifically, I don’t want this TodoOperation to be show to users in the gnarly delimited format (which could mean a difference between Debug and Display?).

Am I mis understanding something?

Reply:
Interesting. When you say "for storage", you mean in the files written to disk, right? I would still be perfectly comfortable treating that as "user-facing output" in that sense. It's effectively a serialization format, right?

Reply:
Ya, exactly it is a serialized output. I guess what I’m thinking is that I’d want the error messages (which I need to make better) to not use this string fornat if I ended up doing a `println!(“found some thing something {}”, todo)` for better error handling. I’d prefer something closer to what `derive(Debug)` does I guess?

Reply:
Ah, that's fair, then. You might go back to implementing it, perhaps with a name like `serialize`, since that's really what it's doing. At that point, it doesn't need a trait, though; you could just do something like:

```rs
impl TodoOperation {
    fn serialize(&self) -> String {
      // the same basic implementation...
    }
}
```

(Since is not `pub fn` it wouldn't be able to be used outside this module.)

Reply:
Oh, thinking more about it: that will of course allocate, so you could also take a buffer which implements `std::io::Write` and/or `std::fmt::Write`, and do this:

```rust
impl TodoOperation {
    fn serialize(&self, buf: std::io::Write) -> std::io::Result<()> {
        write!(
            buf,
            // ...
        )
    }
}
```

See [the docs for `write!`](http://doc.rust-lang.org/1.56.0/core/macro.write.html) for more!